### PR TITLE
Native label scan store as default

### PR DESCRIPTION
**Pull request by tinwelint**
Contains some changes which hindered a full green build with NativeLabelScanStore as the default label scan store:

- One noticeable change is that import-tool no longer builds the label scan store as part of the import, which affects users by delaying the next startup of the db (where the label scan store automatically gets built), but also making import faster. Read more about why this was done in that specific commit.
- Another noticeable change is the setting for users to override which label scan store to use for the db, `dbms.label_index`. Setting is required and defaults to `NATIVE` Most users will never need to touch this setting though. 
- The presence of NativeLabelScanStore made the InMemoryLabelScanStore (used in testing) redundant and so it was also removed.

This PR depends on https://github.com/neo4j/neo4j/pull/8698 since that PR fixes an issue where a checkpoint in the middle of a copy of the native label scan store may corrupt it. It would be wise to merge that PR first.

cl: [kernel] Storage of labels are generally faster and uses less memory due to a new and native label index. This index is set as the default, but can be changed using `dbms.label_index` setting. No migration is required, but the new label index will be automatically built on the next start-up of a db.

### CODE REVIEW

File: community/neo4j/src/test/java/upgrade/StoreUpgraderTest.java
Comment (MishaDemianenko) [inline community/neo4j/src/test/java/upgrade/StoreUpgraderTest.java]: now we have two `@Before` methods :)

File: community/kernel/src/main/java/org/neo4j/graphdb/factory/GraphDatabaseSettings.java
Comment (MishaDemianenko) [inline community/kernel/src/main/java/org/neo4j/graphdb/factory/GraphDatabaseSettings.java]: static modifier is redundant

File: community/kernel/src/main/java/org/neo4j/graphdb/factory/GraphDatabaseSettings.java
Comment (MishaDemianenko) [inline community/kernel/src/main/java/org/neo4j/graphdb/factory/GraphDatabaseSettings.java:490]: do we really want this setting to be public? why users should have possibility to choose implementation? maybe we can make it internal/unsupported?

File: community/lucene-index/src/main/java/org/neo4j/kernel/api/impl/labelscan/LuceneLabelScanIndexBuilder.java
Comment (MishaDemianenko) [inline community/lucene-index/src/main/java/org/neo4j/kernel/api/impl/labelscan/LuceneLabelScanIndexBuilder.java]: this line probably should invoke` buildWritable()` now instead

File: community/lucene-index/src/main/java/org/neo4j/kernel/api/impl/labelscan/LuceneLabelScanStore.java
Comment (MishaDemianenko) [inline community/lucene-index/src/main/java/org/neo4j/kernel/api/impl/labelscan/LuceneLabelScanStore.java]: debug output

File: community/lucene-index/src/main/java/org/neo4j/kernel/api/impl/labelscan/LuceneLabelScanStore.java
Comment (MishaDemianenko) [inline community/lucene-index/src/main/java/org/neo4j/kernel/api/impl/labelscan/LuceneLabelScanStore.java:39]: maybe we should rename this field now :)

File: community/lucene-index/src/main/java/org/neo4j/kernel/api/impl/labelscan/LuceneLabelScanStore.java
Comment (MishaDemianenko) [inline community/lucene-index/src/main/java/org/neo4j/kernel/api/impl/labelscan/LuceneLabelScanStore.java:119]: The whole idea before was to have index that are fully responsible for creation, flushing, closing etc to avoid index volatile variable and necessity of multiple index builders and state handling. With new way it's gone and we need to handle possible multiple indexes and always use volatile variable to read something from index. Maybe we can avoid doing that and come back to old way of doing things?

**Review by MishaDemianenko — CHANGES_REQUESTED** [review]
Couple of comments, otherwise looks great!

File: community/kernel/src/main/java/org/neo4j/graphdb/factory/GraphDatabaseSettings.java
Comment (tinwelint) [reply community/kernel/src/main/java/org/neo4j/graphdb/factory/GraphDatabaseSettings.java:490]: Typically users won't need to care since it will "just work", but at this point in time it's slightly optimistic. Perhaps we could still have it as `@Internal` (which makes it not show up publically, right?)

File: community/lucene-index/src/main/java/org/neo4j/kernel/api/impl/labelscan/LuceneLabelScanStore.java
Comment (tinwelint) [reply community/lucene-index/src/main/java/org/neo4j/kernel/api/impl/labelscan/LuceneLabelScanStore.java:119]: I'm not sure how to do that though

File: community/lucene-index/src/main/java/org/neo4j/kernel/api/impl/labelscan/LuceneLabelScanStore.java
Comment (tinwelint) [reply community/lucene-index/src/main/java/org/neo4j/kernel/api/impl/labelscan/LuceneLabelScanStore.java:39]: Uhm, sure... to what and why?

Comment (MishaDemianenko) [reply community/lucene-index/src/main/java/org/neo4j/kernel/api/impl/labelscan/LuceneLabelScanStore.java:39]: it's not a lucene index anymore right? maybe just to `index` or `labeIndex`?

File: community/kernel/src/main/java/org/neo4j/graphdb/factory/GraphDatabaseSettings.java
Comment (MishaDemianenko) [reply community/kernel/src/main/java/org/neo4j/graphdb/factory/GraphDatabaseSettings.java:490]: yes i think so, vote for `@Internal`

File: community/lucene-index/src/main/java/org/neo4j/kernel/api/impl/labelscan/LuceneLabelScanStore.java
Comment (MishaDemianenko) [reply community/lucene-index/src/main/java/org/neo4j/kernel/api/impl/labelscan/LuceneLabelScanStore.java:119]: We currently already have `ReadOnlyAbstractDatabaseIndex/WritableAbstractDatabaseIndex` but those expect AbstractLuceneIndex maybe those can be updated to expect something else - but those are really lucene oriented currently so that can be some additional work; 
or instead we can just delegate creation to index always - in that case index will replace it's internal structures, not sure how that will work with current new index

File: community/lucene-index/src/main/java/org/neo4j/kernel/api/impl/labelscan/LuceneLabelScanStore.java
Comment (tinwelint) [reply community/lucene-index/src/main/java/org/neo4j/kernel/api/impl/labelscan/LuceneLabelScanStore.java:39]: It's a field in the `LuceneLabelScanStore` so I hope it has to do with lucene :)

File: community/lucene-index/src/main/java/org/neo4j/kernel/api/impl/labelscan/LuceneLabelScanStore.java
Comment (tinwelint) [reply community/lucene-index/src/main/java/org/neo4j/kernel/api/impl/labelscan/LuceneLabelScanStore.java:119]: It'll just be a volatile reference somewhere else, won't it?

Reading a volatile which no one ever writes to, except ones at startup, aint so bad (knowing details about the java memory model), is it?

File: community/lucene-index/src/main/java/org/neo4j/kernel/api/impl/labelscan/LuceneLabelScanStore.java
Comment (MishaDemianenko) [reply community/lucene-index/src/main/java/org/neo4j/kernel/api/impl/labelscan/LuceneLabelScanStore.java:39]: aaaaaaa, true :)

File: community/lucene-index/src/main/java/org/neo4j/kernel/api/impl/labelscan/LuceneLabelScanStore.java
Comment (MishaDemianenko) [reply community/lucene-index/src/main/java/org/neo4j/kernel/api/impl/labelscan/LuceneLabelScanStore.java:119]: yeah thats true, in previous indexes we have a partitions couple of levels below that where guarded anyway;
ok then it lets keep it for now like it is know; that's true - do not think we will have any problems with it

File: community/kernel/src/main/java/org/neo4j/kernel/extension/dependency/NamedLabelScanStoreSelectionStrategy.java
Comment (burqen) [inline community/kernel/src/main/java/org/neo4j/kernel/extension/dependency/NamedLabelScanStoreSelectionStrategy.java]: This javadoc does not look correct anymore. As I understand it, an exception will be thrown if the setting is left unspecified and priority will never be taken into account. Correct?

File: community/kernel/src/main/java/org/neo4j/kernel/impl/api/scan/NativeLabelScanStoreExtension.java
Comment (burqen) [inline community/kernel/src/main/java/org/neo4j/kernel/impl/api/scan/NativeLabelScanStoreExtension.java]: If I've understood it correctly priority is never taken into account when picking `LabelScanStore` and so priority could be removed from both Native and Lucene `...LabelScanStoreExtension`.

File: community/kernel/src/test/java/org/neo4j/kernel/impl/index/labelscan/NativeLabelScanStoreChaosIT.java
Comment (burqen) [inline community/kernel/src/test/java/org/neo4j/kernel/impl/index/labelscan/NativeLabelScanStoreChaosIT.java]: Would be nice to use the `LabelIndex` enum from `GraphDatabaseSettings` explicitly.

File: community/kernel/src/test/java/org/neo4j/kernel/impl/index/labelscan/NativeLabelScanStoreStartupIT.java
Comment (burqen) [inline community/kernel/src/test/java/org/neo4j/kernel/impl/index/labelscan/NativeLabelScanStoreStartupIT.java]: `LabelIndex` enum

File: community/kernel/src/test/java/org/neo4j/kernel/impl/index/labelscan/NativeLabelScanStoreUpdateIT.java
Comment (burqen) [inline community/kernel/src/test/java/org/neo4j/kernel/impl/index/labelscan/NativeLabelScanStoreUpdateIT.java]: `LabelIndex` enum

File: community/lucene-index/src/test/java/org/neo4j/graphdb/LuceneLabelScanStoreChaosIT.java
Comment (burqen) [inline community/lucene-index/src/test/java/org/neo4j/graphdb/LuceneLabelScanStoreChaosIT.java]: `LabelIndex` enum

File: community/lucene-index/src/test/java/org/neo4j/graphdb/LuceneLabelScanStoreStartupIT.java
Comment (burqen) [inline community/lucene-index/src/test/java/org/neo4j/graphdb/LuceneLabelScanStoreStartupIT.java]: `LabelIndex` enum

File: community/lucene-index/src/test/java/org/neo4j/graphdb/LuceneLabelScanStoreUpdateIT.java
Comment (burqen) [inline community/lucene-index/src/test/java/org/neo4j/graphdb/LuceneLabelScanStoreUpdateIT.java]: `LabelIndex` enum

File: community/lucene-index/src/test/java/org/neo4j/index/recovery/UniqueIndexRecoveryTests.java
Comment (burqen) [inline community/lucene-index/src/test/java/org/neo4j/index/recovery/UniqueIndexRecoveryTests.java:105]: Should this test also work with `Native`? It looks like it. Maybe make it abstract and have one implementation for `Lucene` and one for `Native`, like with the `LabelScanStore` specific tests.

**Review by burqen — CHANGES_REQUESTED** [review]
Looks good in general, got some small comments.

File: community/kernel/src/main/java/org/neo4j/kernel/impl/api/scan/NativeLabelScanStoreExtension.java
Comment (tinwelint) [reply community/kernel/src/main/java/org/neo4j/kernel/impl/api/scan/NativeLabelScanStoreExtension.java]: True, we can probably remove this altogether. I'll try that!

File: community/lucene-index/src/test/java/org/neo4j/index/recovery/UniqueIndexRecoveryTests.java
Comment (tinwelint) [reply community/lucene-index/src/test/java/org/neo4j/index/recovery/UniqueIndexRecoveryTests.java:105]: That's really a test of recovery of indexes and uniqueness constraints for the lucene schema index provider. Adding a native variant wouldn't really add anything. It's however unclear why the lucene label index is used specifically, but doesn't seem to matter.

**Review by burqen — APPROVED** [review]

**Comment by burqen** [comment]
@tinwelint Approved! I'll wait for feedback from @MishaDemianenko and CI before merging.

**Pull request merged**
